[Core] Improve module creation documentation - Issue #509 (#623)
pokt-network/pocket: shared/modules/doc/README.md

# Modules <!-- omit in toc -->

> This document outlines how we structured the code splitting it into modules, what a module is and how to create one.

Reviewer: I the first part of this sentence could be more readable by adding the word `by`.

```suggestion
This document outlines how we structured the code by splitting it into modules, what a module is and how to create one.
```

## Definitions

### **Module common interface**

A module common interface is an interface that defines the methods that we modelled as common to multiple modules. For example: the ability to start/stop a module is pretty common and for that we defined the `InterruptableModule` interface.
There are some interfaces that are common to multiple modules and we followed the [Interface segregation principle](https://en.wikipedia.org/wiki/Interface_segregation_principle) and also [Rob Pike's Go Proverb](https://youtu.be/PAAkCSZUG1c?t=317):

> The bigger the interface, the weaker the abstraction.

These interfaces that can be embedded in modules are defined in `shared/modules/module.go`. GoDoc comments will provide you with more information about each interface.

### **Module**

A module is an abstraction (go interface) of a self-contained unit of functionality that carries out a specific task/set of tasks with the idea of being reusable, modular, testable and having a clear and concise API. A module might implement 0..N common interfaces. You can find additional details in the [Modules in detail](#modules-in-detail) section below.

### **Module mock**

> A mock is a fake implementation of a module that is used for testing purposes. It is used to simulate the behaviour of the module and to verify that the module is interacting with other modules correctly. Mocks are generated using `go:generate` directives together with the `mockgen` tool.

Reviewer: I  think the existing description is great, I took a quick look at how [how the wikipedia article describes mocks](https://en.wikipedia.org/wiki/Mock_object) and thought it could be useful to add a bit more detail / rephrasing given the ambiguity of the labels on the mock object spectrum.

```suggestion
A mock is a stand-in, a fake or simplified implementation of a module that is used for testing purposes. It is used to simulate the behaviour of the module and to verify that the module is interacting with other modules correctly. Mocks are generated using `go:generate` directives together with the `mockgen` tool.
```

Reviewer: Wdyt about also making "`mockgen` tool" a link to the `gomock` docs? The README features a "Running mockgen" section above the fold: https://pkg.go.dev/github.com/golang/mock#section-readme

### **Base module**

> A base module is a module that implements a common interface exposing the most basic logic. They are meant to be embedded in the structs of the modules that implement the common interface when there's no special logic to be implemented and improve DRYness (Don't Repeat Yourself) and to reduce boilerplate code. You can find the base modules in the `shared/modules/base_modules` package.

Reviewer: Proposed changes:
1. Replacing the pronoun "they" in the second sentence with "base modules" because it ("they") would be referencing the noun phrase "base module" from the previous sentence which has a different plurality.
1. Split and rephrase end of second sentence.
1. Emboldening the word "embedded" and "and" to highlight the applicability of any given base module.

```suggestion
A base module is a module that implements a common interface, exposing the most basic logic. Base modules are meant to be **embedded** in module structs which implement this common interface **and** don't need to override the respective interface member(s). The intention being to improve DRYness (Don't Repeat Yourself) and to reduce boilerplate code. You can find the base modules in the `shared/modules/base_modules` package.
```

Author: Thank you for your patience @bryanchriswhite! Also for taking the time to explain the proposed change 🙏

# Modules <!-- omit in toc -->

This document outlines how we structured the code by splitting it into modules, what a module is and how to create one.

## Contents <!-- omit in toc -->

- [Definitions](#definitions)
	- [**Module common interface**](#module-common-interface)
	- [**Module**](#module)
	- [**Module mock**](#module-mock)
	- [**Base module**](#base-module)
- [Code Organization](#code-organization)
- [Modules in detail](#modules-in-detail)
	- [Module creation](#module-creation)
	- [Interacting \& Registering with the `bus`](#interacting--registering-with-the-bus)
		- [Modules Registry](#modules-registry)
	- [Start the module](#start-the-module)
	- [Add a logger to the module](#add-a-logger-to-the-module)
	- [Get the module `bus`](#get-the-module-bus)
	- [Stop the module](#stop-the-module)

## Definitions

> ### **Module common interface**

Author: But:

<strong><b>me</b></strong>

:)

> A module common interface is an interface that defines the methods that we modelled as common to multiple modules. For example: the ability to start/stop a module is pretty common and for that we defined the `InterruptableModule` interface.

Reviewer 2: NIT: "shared module interfaces" sounds more intuitive, but maybe you were intending "module common interface" for a reason?

> There are some interfaces that are common to multiple modules and we followed the [Interface segregation principle](https://en.wikipedia.org/wiki/Interface_segregation_principle) and also [Rob Pike's Go Proverb](https://youtu.be/PAAkCSZUG1c?t=317):

Reviewer 2: > In the field of [software engineering](https://en.wikipedia.org/wiki/Software_engineering), the interface segregation principle (ISP) states that no code should be forced to depend on [methods](https://en.wikipedia.org/wiki/Method_(computer_programming)) it does not use

Sounds SOLID

> https://youtu.be/PAAkCSZUG1c?t=317

I guess size does matter 🍆

### **Module mock**

A mock is a stand-in, a fake or simplified implementation of a module that is used for testing purposes. It is used to simulate the behaviour of the module and to verify that the module is interacting with other modules correctly. Mocks are generated using `go:generate` directives together with the [`mockgen` tool](https://pkg.go.dev/github.com/golang/mock#readme-running-mockgen).

### **Base module**

A base module is a module that implements a common interface, exposing the most basic logic. Base modules are meant to be **embedded** in module structs which implement this common interface **and** don't need to override the respective interface member(s). The intention being to improve DRYness (Don't Repeat Yourself) and to reduce boilerplate code. You can find the base modules in the `shared/modules/base_modules` package.

## Code Organization

```bash
├── base_modules            # All the base modules are defined here

> ├── doc			            # Documentation for the modules

Reviewer 2: NIT: Replace tabs with spaces.

Can't believe I'm saying this... https://www.youtube.com/watch?v=SsoOG6ZeyUI

Author: eye_twitches.gif

mad_men_meditate_explosion.gif

I am not gonna start a fight with you **now**  😉

Reviewer 2: At least it would be over a good reason

}
```

For an example of a module that uses a `ModuleOption`, you can search for `WithCustomRPCURL` within the codebase. The code might have changed since this document was written so we are referring to the commit hash [19bf4d3f](https://github.com/pokt-network/pocket/tree/19bf4d3f6507f5d406d9fafdb69b81359bccf110).

Essentially the `ModuleOption` sets a custom RPC URL for the module at runtime.

#### Interacting & Registering with the `bus`

The `bus` is the specific integration mechanism that enables the greater application.

When a module is constructed via the `Create(bus modules.Bus, options ...modules.ModuleOption)` function, it is expected to internally call `bus.RegisterModule(module)`, which registers the module with the `bus` so its sibling modules can access it synchronously via a DI-like pattern.

> Probably worth mentioning that we implemented a `ModulesRegistry` module that takes care of the module registration and retrieval. This module is registered with the `bus` at the application level, it is accessible to all modules via the `bus` interface and it's also mockable as you would expect.

Reviewer: This seems like something that should be represented in the table of contents. Wdyt about 1) giving this paragraph its own heading (e.g. "Module registry" or something) and 2) maybe linking to or embedding an example of registration and/or retrieval?

Author: Totally, I have updated accordingly the TOC and added some color about the modules registry

https://github.com/pokt-network/pocket/pull/623/commits/12a71fa6ab56a6158133405d5759311f11a51455

Reviewer: Amazing -  thank you! :raised_hands:

}
```

> For an example of a module that uses a `ModuleOption`, you can search for `WithCustomRPCURL` within the codebase. The code might have changed since this document was written so we are referring to the commit hash [19bf4d3f](https://github.com/pokt-network/pocket/tree/19bf4d3f6507f5d406d9fafdb69b81359bccf110).

Reviewer 2: 💯

Essentially the `ModuleOption` sets a custom RPC URL for the module at runtime.

### Interacting & Registering with the `bus`

The `bus` is the specific integration mechanism that enables the greater application.

When a module is constructed via the `Create(bus modules.Bus, options ...modules.ModuleOption)` function, it is expected to internally call `bus.RegisterModule(module)`, which registers the module with the `bus` so its sibling modules can access it synchronously via a DI-like pattern.

#### Modules Registry

We implemented a `ModulesRegistry` module [here](https://github.com/pokt-network/pocket/blob/19bf4d3f6507f5d406d9fafdb69b81359bccf110/runtime/modules_registry.go) that takes care of the module registration and retrieval.

> This module is registered with the `bus` at the application level, it is accessible to all modules via the `bus` interface and it's also mockable as you would expect.

Reviewer 2: ```suggestion

This module is registered with the `bus` at the application level, it is accessible to all modules via the `bus` interface and it's also mockable as you would expect.
```

What the `bus` does is setting its reference to the module instance and delegating the registration to the `ModulesRegistry`.

```golang
func (m *bus) RegisterModule(module modules.Module) {
	module.SetBus(m)
	m.modulesRegistry.RegisterModule(module)
}
```

Under the hood, the module name is used to map to the instance of the module so that it's possible to retrieve a module by its name.

This is quite **important** because it unlock a powerful concept **Dependency Injection**.

Essentially this enables the developer to define different implementations of a module and to register the one that is needed at runtime. This is because we can only have one module registered with a unique name and also because, by convention, we keep module names defined as constants.
This is useful not only for prototyping but also for different use cases such as the `p1` CLI and the `pocket` binary where different implementations of the same module are necessary due to the fact that the `p1` CLI doesn't have a persistence module but still needs to know what's going on in the network.

> 

Reviewer 2: ##### Modules Registry Example

> The `peerstore_provider` ([here](https://github.com/pokt-network/pocket/tree/19bf4d3f6507f5d406d9fafdb69b81359bccf110/p2p/providers/peerstore_provider)) is a perfect example of this.

Reviewer 2: ```suggestion
For example, see the `peerstore_provider` ([here](https://github.com/pokt-network/pocket/tree/19bf4d3f6507f5d406d9fafdb69b81359bccf110/p2p/providers/peerstore_provider)).
```

##### Add a logger to the module

> <!-- DISCUSS: I believe we should change this convention, to me it's more semantic if logging is configured during construction/initialization and not during `Start` (which some modules might not even have if they don't implement `InterruptableModule`). I believe that a better approach is summarized here: https://github.com/pokt-network/pocket/blob/8bee148f3b0e768154be4bce02e94813c9382aac/state_machine/module.go#L29-L32 -->

Author: I think this is pretty important but it could be a small separate issue potentially, food for thought.

> <!-- DISCUSS: I believe we should change this convention, to me it's more semantic if logging is configured during construction/initialization and not during `Start` (which some modules might not even have if they don't implement `InterruptableModule`). I believe that a better approach is summarized here: https://github.com/pokt-network/pocket/blob/8bee148f3b0e768154be4bce02e94813c9382aac/state_machine/module.go#L29-L32 -->

Reviewer: Thanks for calling this out @deblasis! :raised_hands: 

It may be slightly tangential but I noticed another seemingly conventional thing that we may want to consider changing:

It looks to me like our `#Create()` **methods** (i.e. not the static pkg-level constructor) return a **new** pointer, essentially ignoring "self". As long as this method's usage is limited to scenarios where the reference to the original receiver of `#Create()` is not shared.

I think my observation can be reduced to, do we expect the following to be true (i.e. pass), because right now it's not:

```golang
package state_machine

import (
	"testing"

	"github.com/golang/mock/gomock"
	"github.com/stretchr/testify/require"

	mockModules "github.com/pokt-network/pocket/shared/modules/mocks"
)

func TestStateMachineModule_Create(t *testing.T) {
	ctrl := gomock.NewController(t)

	busMock := mockModules.NewMockBus(ctrl)
	busMock.EXPECT().PublishEventToBus(gomock.Any()).AnyTimes()
	busMock.EXPECT().RegisterModule(gomock.Any()).AnyTimes()

	mod1 := new(stateMachineModule)

	mod2, err := mod1.Create(busMock)
	require.NoError(t, err)
	require.Equal(t, mod1, mod2)
}
```

To make this true we would need to change our convention in `#Create()` like so:

```
diff --git i/state_machine/module.go w/state_machine/module.go
index b8655084..38bb22e3 100644
--- i/state_machine/module.go
+++ w/state_machine/module.go
@@ -25,8 +25,8 @@ func Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, e
        return new(stateMachineModule).Create(bus, options...)
 }
 
-func (*stateMachineModule) Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
-       m := &stateMachineModule{
+func (m *stateMachineModule) Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
+       *m = stateMachineModule{
                logger: logger.Global.CreateLoggerForModule(modules.StateMachineModuleName),
        }
 
```

> <!-- DISCUSS: I believe we should change this convention, to me it's more semantic if logging is configured during construction/initialization and not during `Start` (which some modules might not even have if they don't implement `InterruptableModule`). I believe that a better approach is summarized here: https://github.com/pokt-network/pocket/blob/8bee148f3b0e768154be4bce02e94813c9382aac/state_machine/module.go#L29-L32 -->

Author: You raise a fair point @bryanchriswhite, I think this should be a separate issue since it doesn't change the signatures of the interfaces.

The idea was that modules should only be instantiated/constructed/initialized from outside their package using the `<package_name>.Create(...)`

This is because we wanted a consistent signature for modules enforced by interfaces but also we wanted to keep the structs private/unexported.

Consequently, this is not a problem because it's not possible to call `new(stateMachineModule).Create(...)` from outside the package since `stateMachineModule` is inaccessible.

To answer your question more directly: I personally don't expect that test to succeed. As in, that's not what I would test for. 
I would test for the fact that the `Create(bus)` sets a bus internally for example.

The `Create` is like a factory on the type and by definition, it "spits out" new instances if that makes sense.

If this sounds confusing even just an inch, let's change this.

> <!-- DISCUSS: I believe we should change this convention, to me it's more semantic if logging is configured during construction/initialization and not during `Start` (which some modules might not even have if they don't implement `InterruptableModule`). I believe that a better approach is summarized here: https://github.com/pokt-network/pocket/blob/8bee148f3b0e768154be4bce02e94813c9382aac/state_machine/module.go#L29-L32 -->

Author: @Olshansk it's an option but then it means that you can't enforce a `create(...)` signature via the interface.

two_red_buttons_cannot_choose_sweating.gif

> <!-- DISCUSS: I believe we should change this convention, to me it's more semantic if logging is configured during construction/initialization and not during `Start` (which some modules might not even have if they don't implement `InterruptableModule`). I believe that a better approach is summarized here: https://github.com/pokt-network/pocket/blob/8bee148f3b0e768154be4bce02e94813c9382aac/state_machine/module.go#L29-L32 -->

Reviewer 2: As a stoic, I pick neither ;)

> <!-- DISCUSS: I believe we should change this convention, to me it's more semantic if logging is configured during construction/initialization and not during `Start` (which some modules might not even have if they don't implement `InterruptableModule`). I believe that a better approach is summarized here: https://github.com/pokt-network/pocket/blob/8bee148f3b0e768154be4bce02e94813c9382aac/state_machine/module.go#L29-L32 -->

Author: 💀

> <!-- DISCUSS: I believe we should change this convention, to me it's more semantic if logging is configured during construction/initialization and not during `Start` (which some modules might not even have if they don't implement `InterruptableModule`). I believe that a better approach is summarized here: https://github.com/pokt-network/pocket/blob/8bee148f3b0e768154be4bce02e94813c9382aac/state_machine/module.go#L29-L32 -->

Reviewer 2: @bryanchriswhite I captured the discussion in #642
